feat: add blog post: Server-side SDKs (#260)
open-feature/openfeature.dev: blog/2023-11-01-feature-flags-sdks-architectures.md

Feature flag evaluation is done on proxy/service only.  
In this approach, as each operation like feature flag evaluation causing network traffic, having a relay proxy in the SDK application cluster can help significantly.

### "Direct" API endpoints Bridge SDK - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Direct-API-endpoints-Bridge-SDK.png').default} />

### "Direct" API endpoints Bridge SDK using Relay Proxy - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Direct-API-endpoints-Bridge-SDK-using-Relay-Proxy.png').default} />

#### Advantages

* Can be relatively simple to develop and maintain from the APIs in multiple code languages, even by code generation.
  The evaluation code can reside at the Relay Proxy, without a need to implement evaluation per SDK language.

#### Disadvantages

* Functional disadvantage is that intermediate issue like network or feature flags service error is handled by not evaluating to the expected feature flag value.

> * Another functional disadvantage is that intermediate issue like network or feature flags service error during microservice initialization is handled by not evaluating to the expected feature flag value.

Author: It is important to understand this point, if it is not clear, let's try to edit or rephrase.
Consider the following flow:
Application starts while the feature flag service connectivity is down for a few minutes.  
At this point, all evaluations are evaluated to the default value, and not the expected value.
Agreed ?

For local evaluation architecture, there is a possible solution for that: the last received configuration from previous start can be persisted locally and used to evaluate to the last expected values.
It is not commonly implemented, but can have an advantage, it is sometimes implemented at the relay proxy.

It is kind of implementation dependent so did not want to dive further for it.
What do you think ?

> * Another functional disadvantage is that intermediate issue like network or feature flags service error during microservice initialization is handled by not evaluating to the expected feature flag value.

Reviewer: Yes, I definately agree that network and service outages can lead to unexpected behavior. However, isn't that already covered in the first point?

> * Another functional disadvantage is that intermediate issue like network or feature flags service error during microservice initialization is handled by not evaluating to the expected feature flag value.

Author: ok, removing.

* Performance and resources usage: each operation like feature flag evaluation causing network traffic.
    * Time-consuming.
    * High load on network traffic.

>     * Affecting both application, and Feature Flag service itself, whether it is self-hosted relay proxy or even more when it is a Cloud Saas service.

Author: "More network requests causing high load on the feature flag service / proxy"
what do you think ?

### API endpoints requests with cache

With this approach, the feature flag service exposes APIs via endpoints like REST HTTP and/or gRpc, and the SDK calling the endpoint requests, and translate the request/responses to the programming language. A cache is used for responses data.  
Feature flag evaluation is done on proxy / features service only.  
As this approach involves network traffic for non-cached evaluations, a Relay Proxy is effective for having the network traffic data on internal network for faster and efficient performance and reducing the load on the Features Service / Source.

### API endpoints requests with cache - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/API-endpoints-requests-with-cache.png').default} />

### API endpoints requests with cache using Relay Proxy - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/API-endpoints-requests-with-cache-using-Relay-Proxy.png').default} />

#### Advantages

> * Network traffic can be reduced.

Reviewer: ```suggestion
* Same advantages as the "Direct" API endpoints bridge architecture.
* Network traffic can be reduced.
```

#### Disadvantages

* Functional disadvantage is that intermediate issue like network or feature flags service error for non-cached data is handled by not evaluating to the expected feature flag value.

> * Another functional disadvantage is that intermediate issue like network or feature flags service error during microservice initialization is handled by not evaluating to the expected feature flag value.

Reviewer: Please see my comments in the `"Direct" API endpoints bridge` section. We should keep these points in line.

* First requests are not in cache, thus slower and causing network traffic.
* Not every request can be efficiently cached, as some evaluations can be dynamic according to context with different values.
* When there is no background task for getting real-time configuration changes event and refreshing the cache, values can take time to be updated.

### Local evaluation

> With this approach, the feature flag configuration is saved locally at the SDK.  

Reviewer: ```suggestion
With this approach, the feature flag configuration is saved locally in the SDK.
```

> Feature flag evaluation is done locally via the SDK and not affecting network traffic.

Reviewer: ```suggestion
Feature flag evaluation is done locally via the SDK and is not affected by network performance.
```

With this approach, configuration can be fetched on initialization, and be refreshed periodically and/or when there is a configuration change, which can be triggered at the SDK via stream event from the feature flag service. 
Some SDKs sync data via WebSockets or [SSE](https://en.wikipedia.org/wiki/Server-sent_events).

> When this approach is used, evaluation analytics metrics can be aggregated at the SDK side and published separately, as not every flag evaluation triggers a message to the Features Service / Proxy.

Reviewer: I would recommend leaving out this sentence since you mentioned it was out of scope in the notes.

### Local evaluation - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Local-evaluation.png').default} />

#### Advantages

> * Functional advantage is that intermediate issue like network or feature flag service errors are automatically handled, as configuration is cached, and flags are evaluated locally.

Reviewer: ```suggestion
* Intermediate issues like network or feature flag service errors have no effect on flag evaluation, as the configuration is cached, and flags are evaluated locally.
```

  If an issue like network error is ongoing for a long time period, SDK provider state can also be reflected as stale.

> * Another functional advantage, initialization can have a fallback mechanism to dealing with situations like temporary network failure by starting from last received configuration or some other configuration, thus microservice can start with last feature flag configuration.

Reviewer: ```suggestion
* Initialization can have a fallback mechanism for dealing with situations like temporary network failure by starting from the last received configuration or some other configuration. Thus, microservice can start with the last feature flag configuration.
```

> * Feature flag configuration can be cached on CDN.

Reviewer: ```suggestion
* Feature flag configuration can be cached on a CDN.
```

* Feature flag evaluation is done locally via the SDK and not affecting network traffic, hence can be fast and efficient.

> * Working without a relay proxy can still be efficient from SDK point of view, as flags are evaluated locally and network traffic does not block common flow.

Reviewer: ```suggestion
* Working without a relay proxy can still be efficient from an SDK point of view, as flags are evaluated locally, and network traffic does not block common flow.
```

#### Disadvantages

> * When feature flag configuration is very large with many feature flags, network traffic and local cache sizes can be increased.

Reviewer: ```suggestion
* When feature flag configurations are very large with many feature flags, network traffic and local cache sizes can be increased. This can impact the time an SDK needs to initialize.
```

* Local evaluation should be implemented at the SDK, developed and maintained for multiple languages. Considering rules like context properties based and percentage rollout strategies via hashing to buckets, it can be hard to keep 100% feature parity across languages.

#### A note on client-side SDKs for web browser / client app

> This approach is less suitable for client-side SDKs from a security point of view, as full feature flag configuration and authorization keys are exposed to end users.

Reviewer: There's some debate on this and some vendors do send flag configurations to the client. I would recommend removing this section for now. Perhaps we could dive into this in more detail in a follow-up blog.

Author: I saw a [reference](https://developer.harness.io/docs/feature-flags/ff-sdks/sdk-overview/client-side-and-server-side-sdks/#:~:text=Server%2Dside%20SDKs%20download%20all,and%20store%20them%20in%20memory.&text=Client%2Dside%20SDK%20has%20a,and%20avoids%20most%20Harness%20communication.) which explains it:
> Security
The SDKs are embedded in your applications and run on the browser or on mobile devices. Therefore, these SDKs can be compromised by users when unpacking a mobile app or using the browser's developer tools to inspect the page. You should not use a server-side SDK key in client-side applications. Client-side SDKs only perform an evaluation of a flag and receive the result. They don't store all data about a flag.

I think it is worth mentioning so readers can at least be aware of some notable difference to client SDKs.
Maybe you want to rephrase it to reduce the debate issue ?

Reviewer: I would recommend avoiding this in general since the blog is specifically focused on server-side SDKs. I do think this is an interesting topic but it requires a deeper dive to truly weigh the pros/cons of each approach. For example, if your rulesets don't contain sensitive information, it may be preferred to send the ruleset instead of sending PII to an external service.

Author: ok, removing it.

> ## OpenFeature and server-side SDKs

Reviewer: Could you add a short introduction to this section? It feels abrupt to jump into a bulleted list.

> * OpenFeature has the [specification](https://openfeature.dev/specification/glossary/#feature-flag-sdk) for SDKs for
>   multiple programming languages for developing a unified API and SDK.

Reviewer: ```suggestion
* OpenFeature has a [specification](https://openfeature.dev/specification/glossary/#feature-flag-sdk) for how an SDK should behave across multiple programming languages.
```

* OpenFeature has a [specification](https://openfeature.dev/specification/glossary/#feature-flag-sdk) for how an SDK should behave across multiple programming languages.
* The OpenFeature specification has methods like flags evaluation and initialization.

> * The OpenFeature specification treats events, which can go together with some scenarios like network errors.

Reviewer: ```suggestion
* The OpenFeature specification supports events, which can be used by developers to react to initialization, flag configuration changes, and error events.
```

> * The OpenFeature specification treats events, which can go together with some scenarios like network errors.

Reviewer: What do you mean by this?

Author: Vendors SDKs commonly expose events like network errors - these events can be mapped to OpenFeature [events](https://openfeature.dev/docs/reference/concepts/events/).

> * As there are multiple SDK architecture approaches, architectures and fallback mechanisms are not always
>   well-defined, and vendors can implement providers the way they choose.

Reviewer: ```suggestion
* As there are multiple SDK architectures, fallback mechanisms are not always well-defined, and vendors can implement providers the way they choose.
```

> 

Reviewer: Could you add a short call to action section? Something about creating a provider and joining the OpenFeature community.

Author: mm not sure it would fit to the post for my point of view.. is that ok ?

Reviewer: Sure, it's up to you.

---
slug: 'feature-flags-sdks-architectures'
title: 'Feature Flags services Server-side SDKs different architectures approaches'
date: 2023-11-01
authors: [liran2000]
description: "Feature Flags services Server-side SDKs different architectures approaches"
tags: [client, sdk]
draft: false
---

Server-side feature flag software development kits (SDKs) are common for letting applications like microservices interact with Feature flag services.

> Feature Flags SDKs have several used functionalities, mainly feature flag evaluation with or without context.

Reviewer 2: change "Flags" to "flag" and remove "used" to read:
Feature flag SDKs have several functionalities, mainly feature flag evaluation with or without context.

> Each Feature Flags service can publish SDKs in multiple programming languages.

Reviewer 2: change "Flags" to "flag"

> The Feature Flags services commonly expose APIs via endpoints like REST HTTP and/or gRPC.

Reviewer 2: change "Flags" to "flag"

> SDKs are an important layer, as network traffic loads affect both the application and the Feature Flags services.

Reviewer 2: change "Feature Flags services" to "feature flag service."

> When the Feature Flags service is a Cloud SaaS service, some services support a relay proxy or a sidecar application, which can be deployed on an organization network.

Reviewer 2: Can we change to read:
When the feature flag service is a Cloud SaaS service, often it supports a relay proxy or a sidecar application, which can be deployed on an organization network.

The Relay Proxy lets multiple clients connect to a local proxy, reducing the number of outbound connections to the Cloud Service.
Considering that a large amount of microservices using SDK instances can be deployed, this can be significant.

<!--truncate-->

## Feature Flags services Server-side SDKs different architectures approaches

During usage and development involvement at several feature flag services providers, I have encountered some different
architectures approaches used by some vendors. Each vendor server-side SDK can commonly belong to one of the described

> approaches.

Reviewer 2: can we change lines 25 - 27 to the following:
Through my own personal use as well as helping to develop several feature flag service providers, I have found that there are a few architectural approaches that most vendors server-side SDKs typically fall under.

Here some of the main server-side SDKs functionalities is described, however, there are additional functionalities like

> statistics and analytics reporting, metrics, events and more.

Reviewer 2: can we change lines 28-29 to read:
Below I dive into three of the main server-side SDK functionalities, but there are additional functionalities like statistics and analytics reporting, metrics, events, and more.

> ### Notes / Glossary

Reviewer 2: remove " / Glossary" so that the heading is just ### Notes

* Described architectures are at high level.
* Features Service described at the diagrams can be a Cloud service, a local network Proxy or some data source
  (some feature flags solutions SDKs can communicate directly from a source like Github).
* There are additional functionalities like statistics reporting, metrics, events and more.

> 

Reviewer 2: Can we change Lines 33 - 35 to read:
* The described architectures below are high level.
* Feature flag services described in the diagrams may be a Cloud service, a local network Proxy or some data source (some feature flags solutions SDKs can communicate directly from a source like Github).

For line 36 - are you saying that these are additional functionalities that you will not cover here? If so, we should make that explicitly clear by saying something to that effect (like below): 
* Additional functionalities like statistics reporting, metrics, events and more should also be taken into consideration, but I have not explored these in this particular blog post.

### "Direct" API endpoints Bridge

> This approach is relatively straight-forward, the Feature Flags service expose APIs via endpoints like

Reviewer 2: change "Flags" to "flag"

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Direct-API-endpoints-Bridge-SDK-using-Relay-Proxy.png').default} />

#### Advantages

* Can be relatively simple to develop and maintain from the APIs in multiple code languages, even by code generation.
  The evaluation code can reside at the Relay Proxy, without a need to implement evaluation per SDK language.

#### Disadvantages

* Functional disadvantage is that intermediate issue like network or feature flags service error is handled by not
  evaluating to the expected feature flag value.
* Another functional disadvantage is that intermediate issue like network or feature flags service error during
  microservice initialization is handled by not evaluating to the expected feature flag value.
* Performance and resources usage: each operation like feature flag evaluation causing network traffic.
    * Time-consuming.
    * High load on network traffic.
    * Affecting both application, and Feature Flag service itself, whether it is self-hosted relay proxy or even more
      when
      it is a Cloud Saas service.

### API endpoints requests with cache

> At this approach, the Feature Flags service expose APIs via endpoints like REST HTTP and/or gRpc, and the SDK calling

Reviewer 2: change "Feature Flags service expose" to "feature flag service exposes"

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/API-endpoints-requests-with-cache.png').default} />

### API endpoints requests with cache using Relay Proxy - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/API-endpoints-requests-with-cache-using-Relay-Proxy.png').default} />

#### Advantages

* Network traffic can be reduced.

#### Disadvantages

* Functional disadvantage is that intermediate issue like network or feature flags service error for non-cached data is
  handled by not evaluating to the expected feature flag value.
* Another functional disadvantage is that intermediate issue like network or feature flags service error during
  microservice
  initialization is handled by not evaluating to the expected feature flag value.
* First requests are not in cache, thus slower and causing network traffic.
* Not every request can be efficiently cached, as some evaluations can be dynamic according to context with different
  values.
* When there is no background task for getting real-time configuration changes event and refreshing the cache,
  values can take time to be updated.

### Local evaluation

> At this approach, the Feature Flags configuration is saved locally at the SDK.  

Reviewer 2: line 110:
Change "At" to "With" and change "Feature Flags" to "feature flag"

> Feature flag evaluation is done on locally via the SDK and not affecting network traffic.

Reviewer 2: remove "on"

With this approach, configuration can be fetched on initialization, and be refreshed periodically and/or when there is

> a configuration change, which can be triggered at the SDK via stream event from the Feature Flags service.

Reviewer 2: change "Feature Flags" to "feature flag"

some SDKs sync data via WebSockets or [SSE](https://en.wikipedia.org/wiki/Server-sent_events).
When this approach is used, evaluation analytics metrics can be aggregated at the SDK side and published separately, as
not every flag evaluation triggers a message to the Features Service / Proxy.

### Local evaluation - example flow

<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Local-evaluation.png').default} />

#### Advantages

> * Functional advantage is that intermediate issue like network or feature flags service error is automatically handled,

Reviewer 2: change "feature flags service error is" to "feature flag service errors are"

  as configuration is cached, and flags are evaluated locally.
  If an issue like network error is ongoing for a long time period, SDK provider state can also be reflected as stale.
* Another functional advantage, initialization can have a fallback mechanism to dealing with situations like temporary
  network failure by starting from last received configuration or some other configuration, thus microservice can start
  with last feature flag configuration.

> * Feature flags configuration can be cached on CDN.

Reviewer 2: change "flags" to "flag"

> * Feature flag evaluation is done on locally via the SDK and not affecting network traffic, hence can be fast and

Reviewer 2: remove "on"

  efficient.
* Working without a relay proxy can still be efficient from SDK point of view, as flags are evaluated locally and
  network traffic does not block common
  flow.

#### Disadvantages

> * When feature flags configuration is very large with many feature flags, network traffic and local cache sizes can be

Reviewer 2: remove s from first "flags" (When feature flag configuration....)

  increased.
* Local evaluation should be implemented at the SDK, developed and maintained for multiple languages. Considering rules
  like context properties based and percentage rollout strategies via hashing to buckets, it can be hard to keep 100%
  feature parity across languages.

#### A note on client-side SDKs for web browser / client app

This approach is less suitable as is for client-side SDKs from security point of view, as full feature flags

> configuration and authorization key are exposed to end users.

Reviewer 2: small grammatical changes, to read:
This approach is less suitable for client-side SDKs from a security point of view, as full feature flag configuration and authorization keys are exposed to end users.

## OpenFeature and server-side SDKs

> * OpenFeature has [specification](https://openfeature.dev/specification/glossary/#feature-flag-sdk) for SDKs for

Reviewer 2: add "the" in OpenFeature has the specification....

  multiple programming languages for developing a unified API and SDK.

> * OpenFeature specification has methods like flags evaluation and initialization.

Reviewer 2: Add "The" to the beginning of this sentence

> * OpenFeature specification treats events, which can go together with some scenarios like network errors.

Reviewer 2: Add "The" to the beginning of this sentence

> * As there are multiple SDKs architectures approaches, architectures and fallback mechanisms are not always

Reviewer 2: change to read "multiple SDK architecture approaches"

---
slug: 'feature-flags-sdks-architectures'
title: 'Feature Flags services Server-side SDKs different architectures approaches'
date: 2023-11-01
authors: [liran2000]
description: "Feature Flags services Server-side SDKs different architectures approaches"
tags: [client, sdk]
draft: false
---

> Server-side SDKs (software development kits) are common for letting applications like microservices interact with
> Feature Flags services.  
> Feature Flags SDKs have several used functionalities, mainly feature flag evaluation with or without a context. Each
> Feature Flags service can publish SDKs in multiple programming languages.  
> The Feature Flags services commonly expose APIs via endpoints like REST HTTP and/or gRpc.
> SDKs are an important layer, as network traffic loads affects both the application and the Feature Flags services.
> When the Feature Flags service is a Cloud SaaS service, some services support a relay proxy or a sidecar application
> which can be deployed on an organization network. The
> Relay Proxy lets multiple clients to connect to a local proxy, reducing number of outbound connections to the Cloud
> Service. Considering that a large amount of microservices using SDKs instances can be deployed, this can be significant.

Reviewer: ```suggestion
Server-side feature flag software development kits (SDKs) are common for letting applications like microservices interact with Feature flag services.
Feature Flags SDKs have several used functionalities, mainly feature flag evaluation with or without context.
Each Feature Flags service can publish SDKs in multiple programming languages.
The Feature Flags services commonly expose APIs via endpoints like REST HTTP and/or gRPC.
SDKs are an important layer, as network traffic loads affect both the application and the Feature Flags services.
When the Feature Flags service is a Cloud SaaS service, some services support a relay proxy or a sidecar application, which can be deployed on an organization network.
The Relay Proxy lets multiple clients connect to a local proxy, reducing the number of outbound connections to the Cloud Service.
Considering that a large amount of microservices using SDK instances can be deployed, this can be significant.
```

<!--truncate-->

> ## Feature Flags services Server-side SDKs different architectures approaches

Reviewer: ```suggestion
## Different approaches for architecture server-side SDKs
```

Author: "Different approaches for server-side SDKs architecture" would be ok ?

Reviewer: Sure, sounds good!

---
slug: 'feature-flags-sdks-architectures'

> title: 'Feature Flags services Server-side SDKs different architectures approaches'

Author: "Different approaches for server-side SDKs architecture" as Michael suggested above would describe it more accurately
